plugin/evm/validators_normalizer_test.go

```diff
@@ -0,0 +1,88 @@
+package evm
+
+import (
+	"math"
+	"math/rand"
+	"reflect"
+	"testing"
+
+	"github.com/flare-foundation/flare/ids"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestValidatorsNormalizer_ByEpoch(t *testing.T) {
+	var (
+		idsList []ids.ShortID
+		weights []uint64
+
+		normTestData     = testData{}
+		copyNormTestData = testData{}
+	)
+
+	numValidators := 6
+	var epochs = [2]uint64{0, 1}
+	for i := 0; i < numValidators; i++ {
+		idsList = append(idsList, randID())
+		weights = append(weights, rand.Uint64())
+
+		if i >= numValidators/len(epochs) {
+			normTestData[epochs[0]] = map[ids.ShortID]uint64{
+				idsList[i]: weights[i],
+			}
+			copyNormTestData[epochs[0]] = map[ids.ShortID]uint64{
+				idsList[i]: weights[i],
+			}
+			continue
+		}
+
+		normTestData[epochs[1]] = map[ids.ShortID]uint64{
+			idsList[i]: weights[i],
+		}
+		copyNormTestData[epochs[1]] = map[ids.ShortID]uint64{
+			idsList[i]: weights[i],
+		}
+	}
+
+	validatorsCache := NewValidatorsCache(normTestData, WithCacheSize(128))
+	for k, v := range normTestData {
+		validatorsCache.cache.Add(k, v)
+	}
+	normalizer := NewValidatorsNormalizer(validatorsCache)
+
+	t.Run("NonExistingKey", func(t *testing.T) {
+		res, err := normalizer.ByEpoch(uint64(7))
+		require.NoError(t, err)
+		assert.Len(t, res, 0)
+	})
+
+	t.Run("ExistingKey", func(t *testing.T) {
+		for _, epoch := range epochs {
+			result, err := normalizer.ByEpoch(epoch)
+			require.NoError(t, err)
+
+			var weightedNTData = map[ids.ShortID]uint64{}
+			for id, _ := range result {
+				weightedNTData[id] = copyNormTestData[epoch][id]
+			}
+
+			weighted := calcWeightRatio(weightedNTData)
+			assert.Equal(t, len(weighted), len(result))
+			assert.True(t, reflect.DeepEqual(weighted, result))
+		}
+	})
+}
+
+func calcWeightRatio(validators map[ids.ShortID]uint64) map[ids.ShortID]uint64 {
+	var totalWeight uint64
+	for _, weight := range validators {
+		totalWeight += weight
+	}
+
+	ratio := math.MaxInt64 / totalWeight
+	for validator, weight := range validators {
+		validators[validator] = weight * ratio
+	}
+
+	return validators
+}
```

Comment:
So here if I get this right, we basically copy the original logic and incorporate it into the test? 😄 I think this is a mistake as it means that if someone modifies the original code they might not realize they need to update it here as well, and also it's not a very DRY approach.

Again I think it would be easier to just hardcode the expected `map[ids.ShortID]uint64` at the top of the file 👍

Reply:
If we do this, then we should also have a test for that extra function, but sure that sounds fine to me. As to not using hardcoded values, I agree when it's simple things but when most of the test consists in defining the variables we use I don't like it 😬 The way we did it in Flow DPS is that we had a [`mocks` package in which we had functions to generate test values randomly](https://github.com/optakt/flow-dps/blob/master/testing/mocks/generic.go) (but also deterministically, so that tests can be reproduced).

WDYT of doing something similar here? Where we have a helper function that generates both the validator set with IDs and weights, and computes the weight ratio for the generated values at the same time?

Reply:
"If we do this, then we should also have a test for that extra function, but sure that sounds fine to me. " - yes, I was thinking the same thing.
Yes, I was thinking about separating the generation into a function last night. It can generate values accordingly for the tests :+1: 
Yes, that will be used for all the tests in validators_ , and not only, tests

Reply:
:handshake:  :smile: